File: jeongsiyeong/python/codetree/binarySearch_02.py
N, M = map(int, input().split())

nums = list(map(int, input().split()))
num_with_cnt = []
cur_num = nums[0]
cur_cnt = 1
for num in nums[1:]:
    if num != cur_num:
        num_with_cnt.append((cur_num, cur_cnt))
        cur_num = num
        cur_cnt = 1
    else:
        cur_cnt += 1
num_with_cnt.append((cur_num, cur_cnt))

def bin_search(m):
    left = 0
    right = len(num_with_cnt)-1
    
    while left <= right:
        mid = (left + right) // 2
        
        if num_with_cnt[mid][0] == m:
            return num_with_cnt[mid][1]
        elif num_with_cnt[mid][0] < m:
            left = mid + 1
        else:
            right = mid - 1
    return 0

for _ in range(M):
    m = int(input())
    print(bin_search(m))

File: jeongsiyeong/cpp/codetree/binarySearch_02.cpp
#include<iostream>
#include<vector>
using namespace std;

int N, M;
vector<pair<int, int>> number_with_cnt;

int bin_search(int m){
  int left = 0;
  int right = number_with_cnt.size() - 1;

  while (left<=right){
    int mid = (left + right) / 2;

    if (number_with_cnt[mid].first == m){
      return number_with_cnt[mid].second;
    }
    else if(number_with_cnt[mid].first < m){
      left = mid + 1;
    }
    else{
      right = mid - 1;
    }
  }
  return 0;
}

int main(){
    cin>>N>>M;

    vector<int> arr;
    for (int i=0; i<N; i++){
      int n;
      cin>>n;
      arr.push_back(n);
    }

    int cur_num=arr[0];
    int cur_cnt=1;

    for(int i=1; i<N; i++){
      if (arr[i] != cur_num){
        number_with_cnt.push_back({cur_num, cur_cnt});
        cur_num = arr[i];
        cur_cnt = 1;
      }
      else{
        cur_cnt++;
      }
    }

    number_with_cnt.push_back({cur_num, cur_cnt});

    for(int i=0; i<M; i++){
      int m;
      cin>>m;
      cout<<bin_search(m)<<"\n";
    }
}